Better implement avatar handling for rooms and occupants.

maranda opened the pull request:
These changes were implemented to improve Bifrost for our usage on the aria-net.org server.

Currently it implements the following features that were missing:
- Fetching of MUC occupant avatars for matrix side of the gateway
- Implementing handling of MUC avatars to set the Matrix room avatar
- Implement logics and facilities to fetch pictures off Matrix profiles and create and store hashes that require to be appended to presences to advertise avatar availability XMPP side of the gateway

Fix a few bugs and issues:
- Attempt fixing mxID unescaping function to better handle multibyte chars
- Fix storage method getRoomsOfType() as it didn't handle correctly anything but group rooms
- Add storage method to retrieve Admin rooms getAdminRoom()
- Fix constraint for admin rooms back to room_id
- Fix pingBridge() as using getRoomByRemoteData() causes it to throw an exception since it'll seek rooms into group_rooms which doesn't have a matrixUser columns, instead use getAdminRoom() with the bot account
- Make sure xmppWriteToStream() always passes a string both when writing to the stream and buffering stanzas, as passing an element now causes promises to break.

maranda committed:
Don't fetch vcards from the MUC itself

it's messy and doesn't do what it is supposed to

maranda committed:
Correct method name

maranda committed:
Use maps instead of sets

maranda committed:
Attempt fixing retrieval of vcard from muc occupants

maranda committed:
Add some verbose logging (temporary)

maranda committed:
Cleanup code

maranda committed:
Merge remote-tracking branch 'upstream/develop' into aria-net-hotfixes

maranda committed:
Attempt fixing mxid to jid mapping

maranda committed:
Merge remote-tracking branch 'upstream/develop' into aria-net-hotfixes

maranda committed:
Attempt computing avatar hashes to add to join presences

maranda committed:
Fix hashing functions to hash binary data

maranda committed:
Change encoding to ascii

maranda committed:
Add avatar hashes also on initial sync'ing

maranda committed:
Fix typo

maranda committed:
Intent is already available in RoomSync

maranda committed:
Don't try to fetch profiles for undefined ghost entries

maranda committed:
Attempt implementing handling of room avatars

maranda committed:
Fix typo

maranda committed:
Set the avatar when we fetch the topic

maranda committed:
Don't try to fetch current avatar when there isn't one

maranda committed:
Fix key name

maranda committed:
Rework previous commits...

- Add new event interface which supports passing buffers
- Try not mangling the file type and make it unusable

maranda committed:
currentData is supposed to be the body

maranda committed:
Don't fetch profiles from not unready ghosts

maranda committed:
Rename IChatBufferState interface

maranda committed:
Merge remote-tracking branch 'upstream/develop' into aria-net

maranda committed:
Fix merging conflict mistake

maranda committed:
Use axios instead of request-promise-native

maranda committed:
We shall not pass the element when buffering messages

maranda committed:
Don't also pass the element to stream write()

maranda committed:
Catch if we're being sent an element

maranda committed:
Add missing hashes to presences

The commit got ate when resolving merge conflicts

maranda committed:
Fix admin rooms

- Add storage function to get admin rooms
- Fix getRoomsOfType() as it didn't actually get correctly anything else but group rooms
- Fix bridge pinging on startup

maranda committed:
Constraint for admin rooms is room_id

maranda committed:
Attempt implementing avatar hashes in xmpp gateway rooms

maranda committed:
Fix asynchronous calls in previous commit

maranda committed:
Improve logging

maranda committed:
Propagate hashes to MUC membership

And emit presences via the XJSGateway.

maranda committed:
Correct parameters

maranda committed:
Remove unused import

maranda committed:
Add changelog to pull request

tadzik commented on src/xmppjs/XJSInstance.ts:
```diff
@@ -117,11 +120,16 @@ export class XmppJsInstance extends EventEmitter implements IBifrostInstance {
     }
 
     public xmppWriteToStream(xmlMsg: any) {
```
I worry that this might hide errors more than fix them – I'd rather give this a stronger type signature and make sure we know what we're dealing with.

Looking through the code though, this seems to always be called either with `Element` (across `ServiceHandler`) or once with a `StzaPresenceItem` (in `XJSGateway.remoteLeft`), at least on current master. Looking at the latter, I'm not sure if `toString()` even turns it into a proper XML string (`xml()` seems to be responsible for that).

How about we make the type checks stronger here, and make sure we're handling things correctly?

maranda replied:
No problem by my side, my change was just to fix broken promises as this.xmpp.write() only accepts either strings or a buffer after the dependencies bump (previously it worked just fine with elements).

tadzik commented on src/xmppjs/XJSInstance.ts:
```diff
@@ -430,13 +438,59 @@ export class XmppJsInstance extends EventEmitter implements IBifrostInstance {
+            const timeout = setTimeout(() => reject(Error("Timeout")), 5000);
+            this.once(`iq.${id}`, (stanza: Element) => {
+                clearTimeout(timeout);
+                const vCard = (stanza.getChild("vCard") as unknown as Element);
+                if (vCard) {
+                    resolve(vCard);
+                }
```
These type casts shouldn't  be necessary – this should already be `Element|undefined`.

```suggestion
                const vCard = stanza.getChild("vCard");
                if (vCard) {
                    resolve(vCard!);
                }
```

maranda replied:
No problem on the change, I just left it as it previously was.

maranda replied:
Actually removing the type casts cause the following when building:
```
src/xmppjs/XJSInstance.ts(498,29): error TS2345: Argument of type 'import("/node_modules/@types/ltx/lib/Element").Element' is not assignable to parameter of type 'import("/node_modules/@types/xmpp__xml/index").Element'.
  Type 'Element' is missing the following properties from type 'Element': append, prepend
```

So I'll revert the change for now.

maranda replied:
@tadzik could you review the type cast issue above? As your suggested change causes the problem above.

tadzik commented on src/xmppjs/XJSInstance.ts:
```diff
@@ -430,13 +438,59 @@ export class XmppJsInstance extends EventEmitter implements IBifrostInstance {
         const whoJid = jid(who);
         who = `${whoJid.local}@${whoJid.domain}`;
-        log.info(`Fetching vCard for ${who}`);
+        let muc = this.checkMUCCache.get(who);
```
```suggestion
        let mucExists = this.checkMUCCache.get(who);
```
Should be a little easier to read the rest of the method :)

maranda replied:
Agreed.

tadzik commented on src/xmppjs/XJSInstance.ts:
```diff
@@ -446,7 +500,8 @@ export class XmppJsInstance extends EventEmitter implements IBifrostInstance {
             });
         });
-        // Remove the resource
+        // Send the request
```
```suggestion
```

maranda replied:
+1

tadzik reviewed (changes requested):
Great work! Left some nitpicks and suggestions.

This seems to be overlapping slightly with https://github.com/matrix-org/matrix-bifrost/pull/272, so I'd rather remove these changes here, since they have their own specialized PR and all.

maranda committed:
Add suggested changes to XJSInstance

- Make it clearer what the method does
- Remove typing casts from the vCard
- Remove comment

maranda commented:
> This seems to be overlapping slightly with #272, so I'd rather remove these changes here, since they have their own specialized PR and all.

That's fine but I'm not sure how to bisect those changes from the rest without remaking a pull request 🤔

maranda committed:
Revert type cast removal as it causes a build error

Half-Shot commented:
> > This seems to be overlapping slightly with #272, so I'd rather remove these changes here, since they have their own specialized PR and all.
> 
> That's fine but I'm not sure how to bisect those changes from the rest without remaking a pull request thinking

I've since merged those changes, so you just need to resolve the merge and I think we can re-review from there.

maranda committed:
Merge remote-tracking branch 'upstream/develop' into aria-net

maranda committed:
Update changelog to reflect merge conflict sorting

maranda committed:
Cache the avatar hash on the account object for retries and reconnects

maranda committed:
Merge remote-tracking branch 'upstream/tadzik/optimize-updateMatrixMemberListForRoom' into aria-net

maranda committed:
Merge branch 'matrix-org:develop' into aria-net

maranda committed:
Add missing avatar hash from merge

maranda committed:
Remove sensitive data from saved properties

maranda committed:
Return modified properties

maranda commented:
Will rebase the pull request, this is becoming a mess to mantain with all syncs and is unreadable.
